File: 08-greedy_algorithms/completed/greedy_sets-COMPLETED.rb
require "rspec/autorun"
require "set"

def greedy_search(states_needed, stations)
    final_stations = []
    # loop until states_needed is empty
    while states_needed.length > 0 do
        best_station = nil
        states_covered = Set.new
        stations.each { |station_name, states_for_station|
            # find what remaining states this station covers
            covered = states_needed & states_for_station
            # if this station covers the most states, overwrite the previous best station
            if covered.length > states_covered.length
                best_station = station_name
                states_covered = covered
            end
        }
        # remove states_covered from states_needed
        states_needed -= states_covered
        # add best_station to final_stations
        final_stations.push(best_station)
    end
    final_stations
end

# tests here
describe 'greedy search' do
    it "tests that the greedy search works" do
        states_needed = Set.new(['mt', 'wa', 'or', 'id', 'nv', 'ut', 'ca', 'az'])

        stations = {
            'kone' => Set.new(['id', 'nv', 'ut']),
            'ktwo' => Set.new(['wa', 'id', 'mt']),
            'kthree' => Set.new(['or', 'nv', 'ca']),
            'kfour' => Set.new(['nv', 'ut']),
            'kfive' => Set.new(['ca', 'az']),
        }
        result = greedy_search(states_needed, stations)
        expect(result).to eq(['kone', 'ktwo', 'kthree', 'kfive'])
    end

    # TODO: add more test cases
end

File: 08-greedy_algorithms/completed/greedy_sets-COMPLETED.js
const testEquality = require('../../testingFunction');

const getIntersection = (set1, set2) => {
    return new Set([...set1].filter(item => set2.has(item)));
}

const getDifference = (set1, set2) => {
    return new Set([...set1].filter(item => !set2.has(item)));
}

const getUnion = (set1, set2) => {
    return new Set([...set1].concat([...set2]));
}

statesNeeded = new Set(['mt', 'wa', 'or', 'id', 'nv', 'ut', 'ca', 'az'])

stations = {
    kone: new Set(['id', 'nv', 'ut']),
    ktwo: new Set(['wa', 'id', 'mt']),
    kthree: new Set(['or', 'nv', 'ca']),
    kfour: new Set(['nv', 'ut']),
    kfive: new Set(['ca', 'az']),
}

finalStations = [];

while ([...statesNeeded].length) {
    bestStation = null;
    statesCovered = new Set();
    Object.keys(stations).forEach(station => {
        statesForStation = stations[station]
        covered = getIntersection(statesNeeded, statesForStation)
        if ([...covered].length > [...statesCovered].length) {
            bestStation = station;
            statesCovered = covered;
        }
    });
    statesNeeded = getDifference(statesNeeded, statesCovered)
    finalStations.push(bestStation)
}

testEquality(finalStations, ['kone', 'ktwo', 'kthree', 'kfive']);
